### PR TITLE
sign_up/sign_in

### PR DESCRIPTION
**Pull request by itsaayush07**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/gorilla/mux"
+)
+
+//===========================================================================================
```
Comment (chirag-applift) [inline main.go]: Remove these comments. 
In most cases, a new line is enough to separate different sections of code. If not, you can always create new function and call it.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+//===========================================================================================
+//MODEL
+var db *sql.DB
+var err error
+
+/*create table users(Username varchar(10) primary,
```
Comment (chirag-applift) [inline main.go]: Create a separate .sql file and keep updating it with sql queries manually, or read about how to manage sql migration in golang.

Remove this comment from here.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+FirstName varchar(10),LastName varchar(10),Password varchar(20),
+Gender varchar(10),Country varchar(20),Age varchar(10)); */
+
+//User struct
+type User struct {
```
Comment (chirag-applift) [inline main.go]: Create a different package called models for structs which represents tables.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	Password  string `json:"password"`
+	Gender    string `json:"gender"`
+	Country   string `json:"country"`
+	Age       string `json:"age"`
+	Email     string `json:"email"`
+}
+
+//===========================================================================================
+//CONTROLLERSb
+func handleSignUp(w http.ResponseWriter, r *http.Request) {
```
Comment (chirag-applift) [inline main.go]: In golang controllers are called handlers, as you may know already. Create a separate package called handlers there.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	// Email := keyVal["email"]
+
+	_, err = stmt.Exec(user.Username, user.FirstName, user.LastName, user.Password, user.Gender, user.Country, user.Age, user.Email)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Print("new user is created.")
+}
+
+// func handleSignIn(w http.ResponseWriter, r *http.Request) {
```
Comment (chirag-applift) [inline main.go]: Don't comment old code. Just commit them, and then remove completely. This keeps the code clean, easy to read and you can checkout old commit to see your old code, if needed.

File: main.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+var db *sql.DB
+var err error
+
+func main() {
+
+	r := mux.NewRouter()
+
+	db, err = sql.Open("mysql", "root:mysql1234@tcp(localhost:3306)/dashboard")
```
Comment (chirag-applift) [inline main.go]: Move this to separate package called db.

File: controller/controller.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+var db *sql.DB
+var err error
+
+//HandleSignUp func
+func HandleSignUp(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	var user model.User
+	stmt, err := db.Prepare("INSERT INTO users(Username,FirstName,LastName,Password,Gender,Country,Age,Email) VALUES(?)")
```
Comment (chirag-applift) [inline controller/controller.go]: Move this to separate package called db.

File: controller/controller.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = json.Unmarshal(body, &user)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	_, err = stmt.Exec(user.Username, user.FirstName, user.LastName, user.Password, user.Gender, user.Country, user.Age, user.Email)
```
Comment (chirag-applift) [inline controller/controller.go]: Move this to separate package called db.

File: controller/controller.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
+package controller
```
Comment (chirag-applift) [inline controller/controller.go]: rename controllers to handlers,that way, it's easier for other Golang developers to understand your code

File: model/model.go [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+	LastName  string `json:"lastname"`
+	Password  string `json:"password"`
+	Gender    string `json:"gender"`
+	Country   string `json:"country"`
+	Age       string `json:"age"`
+	Email     string `json:"email"`
+}
+
+//ResponseResult struct
+type ResponseResult struct {
```
Comment (chirag-applift) [inline model/model.go]: Structs like User are used by handlers package and db package both, that's why we keep them in separate package called model. But ResponseResult struct is specific to handlers package and will not be used elsewhere, so keep  it in handlers package only.

File: controller/controller.go (hunk 7 above)
Comment (itsaayush07) [reply controller/controller.go]: should i move the prepare and exec statements also to db package?

**Comment by chirag-applift** [comment]
All of your commits say "Add files via upload".
It should be something that represent work done in that commit (like "added signup handler", "refactored code" or "added signin handler"). So it's easier for you to find commit related to some feature in future.

File: model/model.go (hunk 10 above)
Comment (itsaayush07) [reply model/model.go]: okay.

**Comment by itsaayush07** [comment]
> All of your commits say "Add files via upload".
> It should be something that represent work done in that commit (like "added signup handler", "refactored code" or "added signin handler"). So it's easier for you to find commit related to some feature in future.

will  take care of that now.

File: db/dashboard_users.sql [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+CREATE TABLE `users` (
+  `Username` varchar(10) NOT NULL,
+  `FirstName` varchar(10) DEFAULT NULL,
+  `LastName` varchar(10) DEFAULT NULL,
+  `Password` varchar(20) DEFAULT NULL,
+  `Gender` varchar(10) DEFAULT NULL,
+  `Country` varchar(20) DEFAULT NULL,
+  `Age` varchar(10) DEFAULT NULL,
+  `Email` varchar(30) DEFAULT NULL,
+  PRIMARY KEY (`Username`)
```
Comment (chirag-applift) [inline db/dashboard_users.sql]: add id column which will be integer and primary key.
Add username and email unique constraint.

Comment (itsaayush07) [reply db/dashboard_users.sql]: okay

File: controller/controller.go [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+	defer stmtOut.Close()
+	err = stmtOut.QueryRow(user.Username).Scan(&query)
+	if err != nil {
+		stmt, err := db.Prepare("INSERT INTO users(Username,FirstName,LastName,Password,Gender,Country,Age,Email) VALUES(?)")
+		if err != nil {
+			log.Panic(err)
+		}
+		defer stmt.Close()
+
+		_, err = stmt.Exec(user.Username, user.FirstName, user.LastName, user.Password, user.Gender, user.Country, user.Age, user.Email)
```
Comment (chirag-applift) [inline controller/controller.go]: do not insert password as plain text in db. Use sha1 library to calculate hash and then store hash in db.
and while sign in again calculate sha1 and match from db.

File: controller/controller.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+		json.NewEncoder(w).Encode(res)
+		return
+	}
+	res.Error = "Already Registered Username"
+	json.NewEncoder(w).Encode(res)
+
+}
+
+//HandleSignIn func
+func HandleSignIn(w http.ResponseWriter, r *http.Request) {
```
Comment (chirag-applift) [inline controller/controller.go]: Create separate branch for signin code and create separate pull request

Comment (itsaayush07) [reply controller/controller.go]: done.

File: controller/controller.go (hunk 7 above)
Comment (chirag-applift) [reply controller/controller.go]: Yes, everything related to db operations (like insert, delete, select should be in db package) 
func insertUser(user *User) {
// prepare and exec.
}
and  you should call the function in handlers package like 
db.insertUser(&user)

Comment (itsaayush07) [reply controller/controller.go]: okay.

**Comment by chirag-applift** [comment]
@itsaayush07  all of your pull requests have the same code (all of it). This is no advantage of multiple pull requests. So, you either create only one pull request and wait for it to merge to master before writing any new feature or you base your 2nd pull request from 1st pull request's branch. Basically there should not be any code that is present in more than one pull request.

**Comment by itsaayush07** [comment]
> @itsaayush07 all of your pull requests have the same code (all of it). This is no advantage of multiple pull requests. So, you either create only one pull request and wait for it to merge to master before writing any new feature or you base your 2nd pull request from 1st pull request's branch. Basically there should not be any code that is present in more than one pull request.

**Pull request closed without merging**